examples/data_process/tutorial_fast_affine_transform.py

```diff
@@ -0,0 +1,131 @@
+"""
+Tutorial of fast affine transformation.
+To run this tutorial, install opencv-python using pip.
+
+Comprehensive explanation of this tutorial can be found https://tensorlayer.readthedocs.io/en/stable/modules/prepro.html
+"""
+
+import multiprocessing
+import time
+
+import tensorflow as tf
+import tensorlayer as tl
+import numpy as np
+import cv2
+
+# tl.logging.set_verbosity(tl.logging.DEBUG)
+image = tl.vis.read_image('tiger.jpeg')
+h, w, _ = image.shape
+
+
+def create_transformation_matrix():
+    # 1. Create required affine transformation matrices
+    M_rotate = tl.prepro.affine_rotation_matrix(angle=20)
+    M_flip = tl.prepro.affine_horizontal_flip_matrix(prob=1)
+    M_shift = tl.prepro.affine_shift_matrix(wrg=0.1, hrg=0, h=h, w=w)
+    M_shear = tl.prepro.affine_shear_matrix(x_shear=0.2, y_shear=0)
+    M_zoom = tl.prepro.affine_zoom_matrix(zoom_range=0.8)
+
+    # 2. Combine matrices
+    # NOTE: operations are applied in a reversed order (i.e., rotation is performed first)
+    M_combined = M_shift.dot(M_zoom).dot(M_shear).dot(M_flip).dot(M_rotate)
+
+    # 3. Convert the matrix from Cartesian coordinates (the origin in the middle of image)
+    # to image coordinates (the origin on the top-left of image)
+    transform_matrix = tl.prepro.transform_matrix_offset_center(M_combined, x=w, y=h)
+    return transform_matrix
+
+
+def example1():
+    """ Example 1: Applying transformation one-by-one is very SLOW ! """
+    st = time.time()
+    for _ in range(100):  # Try 100 times and compute the averaged speed
+        xx = tl.prepro.rotation(image, rg=-20, is_random=False)
+        xx = tl.prepro.flip_axis(xx, axis=1, is_random=False)
+        xx = tl.prepro.shear2(xx, shear=(0., -0.2), is_random=False)
+        xx = tl.prepro.zoom(xx, zoom_range=1 / 0.8)
+        xx = tl.prepro.shift(xx, wrg=-0.1, hrg=0, is_random=False)
+    print("apply transforms one-by-one took %fs for each image" % ((time.time() - st) / 100))
+    tl.vis.save_image(xx, '_result_slow.png')
+
+
+def example2():
+    """ Example 2: Applying all transforms in one is very FAST ! """
+    st = time.time()
+    for _ in range(100):  # Repeat 100 times and compute the averaged speed
+        transform_matrix = create_transformation_matrix()
```

Comment:
should `transform_matrix = create_transformation_matrix()` be outside the for loop?

Reply:
Resolved.